statsmodels/tsa/mstl.py

```diff
@@ -0,0 +1,240 @@
+"""
+from typing import Dict, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from scipy.stats import boxcox
+
+from statsmodels.tools.typing import ArrayLike1D
+from statsmodels.tsa.seasonal import STL
+from statsmodels.tsa.tsatools import freq_to_period
+
+
+class MSTL:
+    """
+    MSTL(endog, periods=None, windows=None, lmbda=None, iterate=2, **kwargs)
+
+    Season-Trend decomposition using LOESS for multiple seasonalities.
+
+    Parameters
+    ----------
+    endog : array_like
+        Data to be decomposed. Must be squeezable to 1-d.
+    periods : {int, array_like, None}, optional
+        Periodicity of the seasonal components. If None and endog is a pandas
+        Series or DataFrame, attempts to determine from endog. If endog is a
+        ndarray, periods must be provided.
+    windows : {int, array_like, None}, optional
+        Length of the seasonal smoothers for each corresponding period.
+        Must be an odd integer, and should normally be >= 7 (default). If None
+        then default values determined using 7 + 4 * np.arange(1, n + 1, 1)
+        where n is number of seasonal components.
+    lmbda : {float, str, None}, optional
+        The lambda parameter for the Box-Cox transform to be applied to `endog`
+        prior to decomposition. If None, no transform is applied. If "auto", a
+        value will be estimated that maximizes the log-likelihood function.
+    iterate : int, optional
+        Number of iterations to use to refine the seasonal component.
+
+    Other Parameters
+    ----------------
+    **kwargs : dict
+        Other arguments are passed to STL.
+
+    See Also
+    --------
+    statsmodels.tsa.seasonal.STL
+
+    References
+    ----------
+    .. [1] K. Bandura, R.J. Hyndman, and C. Bergmeir (2021)
+    MSTL: A Seasonal-Trend Decomposition Algorithm for Time Series with
+    Multiple Seasonal Patterns. arXiv preprint arXiv:2107.13462.
+
+    Examples
+    --------
+    Start by creating a toy dataset with multiple seasonal components.
+
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> register_matplotlib_converters()
+    >>> t = np.arange(1, 1000)
+    >>> trend = 0.0001 * t ** 2 + 100
+    >>> daily_seasonality = 5 * np.sin(2 * np.pi * t / 24)
+    >>> weekly_seasonality = 10 * np.sin(2 * np.pi * t / (24 * 7))
+    >>> noise = np.random.randn(len(t))
+    >>> y = trend + daily_seasonality + weekly_seasonality + noise
+
+    Use MSTL to decompose the time series into multiple seasonal components
+    with periods 24 (daily) and 24*7 (weekly).
+
+    >>> from statsmodels.tsa.seasonal import MSTL
+    >>> res = STL(data, periods=(24, 24*7)).fit()
+    >>> res.plot()
+    >>> plt.show()
+    """
+
+    def __init__(
+        self,
+        endog: ArrayLike1D,
+        periods: Optional[Union[int, Sequence[int]]] = None,
+        windows: Optional[Union[int, Sequence[int]]] = None,
+        lmbda: Optional[Union[float, str]] = None,
```

Comment:
I think many of these should be keyword only.  Not sure where to draw the line.  KW only make it a lot easier to make any future changes.